File: skills/cloudflare-proxy/scripts/proxy_fetch.py
#!/usr/bin/env python3
"""
Cloudflare Worker Proxy — Fetch 模式调用器

注意：由于 macOS Python ssl 模块不支持 Cloudflare Workers 要求的 TLS 版本，
实际抓取由 proxy_fetch.js（Node.js）完成。

用法：
    python3 proxy_fetch.py <目标URL> [--max-chars N] [--proxy <代理URL>]

示例：
    python3 proxy_fetch.py https://github.com
    python3 proxy_fetch.py https://huggingface.co --max-chars 200000
"""

import subprocess
import sys
import os
import shutil

SCRIPT_DIR = os.path.dirname(os.path.abspath(__file__))
JS_SCRIPT = os.path.join(SCRIPT_DIR, "proxy_fetch.js")
DEFAULT_PROXY = "https://arthur-proxy.frankfeahter.workers.dev"


def main():
    args = sys.argv[1:]

    if not args or "--help" in args:
        print(__doc__)
        sys.exit(0)

    # 找到 Node.js
    node = shutil.which("node") or "/usr/local/bin/node"
    if not os.path.exists(node):
        print("ERROR: Node.js not found. Please install Node.js.", file=sys.stderr)
        sys.exit(1)

    # 构建 Node.js 调用参数
    js_args = [node, JS_SCRIPT] + args
    result = subprocess.run(js_args, capture_output=True, text=True)
    sys.stdout.write(result.stdout)
    if result.stderr:
        sys.stderr.write(result.stderr)
    sys.exit(result.returncode)


if __name__ == "__main__":
    main()

File: skills/cloudflare-proxy/scripts/proxy_fetch.js
#!/usr/bin/env node
/**
 * Cloudflare Worker Proxy — Fetch 模式（Node.js 实现）
 *
 * macOS 的 Python ssl 模块不支持 Cloudflare Workers 的 TLS 版本，
 * 所以改用 Node.js（使用自己的 crypto，完美支持 TLS 1.3）。
 */

const https = require('https');

const args = process.argv.slice(2);
const maxChars = 50000;

if (!args[0] || args[0] === '--help') {
  console.log('用法: node proxy_fetch.js <目标URL> [maxChars] [--proxy <代理URL>]');
  console.log('示例: node proxy_fetch.js https://github.com 30000');
  process.exit(0);
}

let targetUrl = args[0];
let proxy = 'https://arthur-proxy.frankfeahter.workers.dev';
let maxCharsOverride = maxChars;

// Parse flags
const remainingArgs = args.slice(1);
for (let i = 0; i < remainingArgs.length; i++) {
  if (remainingArgs[i] === '--proxy' && remainingArgs[i + 1]) {
    proxy = remainingArgs[i + 1];
    i++;
  } else if (!isNaN(parseInt(remainingArgs[i]))) {
    maxCharsOverride = parseInt(remainingArgs[i]);
  }
}

const encodedTarget = encodeURIComponent(targetUrl);
const proxyFetchUrl = `${proxy.replace(/\/$/, '')}/?url=${encodedTarget}&_raw=1`;

const req = https.get(proxyFetchUrl, {
  headers: {
    'User-Agent': 'Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/120.0.0.0 Safari/537.36',
    'Accept': 'text/html,application/xhtml+xml,application/xml;q=0.9,*/*;q=0.8',
    'Accept-Language': 'en-US,en;q=0.9,zh-CN;q=0.8',
    'Connection': 'keep-alive',
  },
  timeout: 30000,
}, (res) => {
  if (res.statusCode >= 400) {
    console.error(`HTTP ${res.statusCode}`);
    process.exit(1);
  }

  const chunks = [];
  res.on('data', chunk => chunks.push(chunk));
  res.on('end', () => {
    let content = Buffer.concat(chunks).toString('utf8');

    // Cloudflare challenge detection
    if (content.includes('cf-challenge') || content.includes('turnstile') ||
        content.includes('_cf_translation') || content.includes('cloudflare')) {
      // Check if it's actually the challenge page
      if (content.includes('Ray ID') || content.includes('的人机验证')) {
        console.error('ERROR: CF验证墙拦截 — 自建 Worker 可避免此问题');
        process.exit(1);
      }
    }

    // Check if we got the homepage instead of target content
    if (content.includes('CF Proxy') && !content.includes(targetUrl)) {
      // We got the proxy homepage — target wasn't fetched correctly
      const errMsg = `Proxy returned homepage instead of target content. URL: ${targetUrl}`;
      console.error('ERROR:', errMsg);
      process.exit(1);
    }

    if (content.length > maxCharsOverride) {
      content = content.slice(0, maxCharsOverride) +
        `\n\n[Output truncated at ${maxCharsOverride} chars, full: ${content.length} chars]`;
    }
    process.stdout.write(content);
  });
});

req.on('error', (e) => {
  console.error('Network error:', e.message);
  process.exit(1);
});

req.on('timeout', () => {
  req.destroy();
  console.error('Request timeout');
  process.exit(1);
});
